src/realm/parser/query_builder.cpp

```diff
@@ -0,0 +1,941 @@
+            break;
+        case Predicate::Operator::NotEqual:
+            query.and_query(column.not_equal(value, case_sensitive));
+            break;
+        case Predicate::Operator::BeginsWith:
+            query.and_query(column.begins_with(value, case_sensitive));
+            break;
+        case Predicate::Operator::EndsWith:
+            query.and_query(column.ends_with(value, case_sensitive));
+            break;
+        case Predicate::Operator::Contains:
+            query.and_query(column.contains(value, case_sensitive));
+            break;
+        case Predicate::Operator::Like:
+            query.and_query(column.like(value, case_sensitive));
+            break;
+        default:
+            throw std::logic_error("Unsupported operator for keypath substring queries.");
+    }
+}
+
+void add_binary_constraint_to_query(Query &query,
+                                    Predicate::Operator op,
+                                    Columns<Binary> &&column,
+                                    BinaryData &&value) {
+    switch (op) {
+        case Predicate::Operator::BeginsWith:
+            query.begins_with(column.column_ndx(), BinaryData(value));
+            break;
+        case Predicate::Operator::EndsWith:
+            query.ends_with(column.column_ndx(), BinaryData(value));
+            break;
+        case Predicate::Operator::Contains:
+            query.contains(column.column_ndx(), BinaryData(value));
+            break;
+        case Predicate::Operator::Equal:
+            query.equal(column.column_ndx(), BinaryData(value));
+            break;
+        case Predicate::Operator::NotEqual:
+            query.not_equal(column.column_ndx(), BinaryData(value));
+            break;
+        default:
+            throw std::logic_error("Unsupported operator for binary queries.");
+    }
+}
+
+void add_binary_constraint_to_query(realm::Query &query,
+                                    Predicate::Operator op,
+                                    BinaryData value,
+                                    Columns<Binary> &&column) {
+    switch (op) {
+        case Predicate::Operator::Equal:
+            query.equal(column.column_ndx(), BinaryData(value));
+            break;
+        case Predicate::Operator::NotEqual:
+            query.not_equal(column.column_ndx(), BinaryData(value));
+            break;
+        default:
+            throw std::logic_error("Substring comparison not supported for keypath substrings.");
+    }
+}
+
+void add_link_constraint_to_query(realm::Query &query,
+                                  Predicate::Operator op,
+                                  const PropertyExpression &prop_expr,
+                                  size_t row_index) {
+    precondition(prop_expr.indexes.empty(), "KeyPath queries not supported for object comparisons.");
+    switch (op) {
+        case Predicate::Operator::NotEqual:
+            query.Not();
+            //REALM_FALLTHROUGH;
```

Comment:
If someone changes this code while completely ignoring the `REALM_FALLTHROUGH` annotation, then yes, it could be problematic. But how likely is it really that someone would miss that annotation?